catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/metacard/MetacardApplication.java

```diff
@@ -534,28 +546,112 @@ public void init() {
-
-          Metacard metacard = transformer.transform(workspace);
-          metacard.setAttribute(new AttributeImpl(Metacard.ID, id));
+          Metacard updated = transformer.transform(body);
+          updated.setAttribute(new AttributeImpl(Core.ID, id));
 
-          Metacard updated = updateMetacard(id, metacard);
-          return util.getJson(transformer.transform(updated));
+          Metacard stored = updateMetacard(id, updated);
+          return util.getJson(transformer.transform(stored));
         });
 
     delete(
         "/workspaces/:id",
         APPLICATION_JSON,
         (req, res) -> {
           String id = req.params(":id");
+          WorkspaceMetacardImpl workspace = workspaceService.getWorkspaceMetacard(id);
+
+          String[] queryIds = workspace.getQueries().toArray(new String[0]);
+
+          if (queryIds.length > 0) {
+            catalogFramework.delete(new DeleteRequestImpl(queryIds));
+          }
+
           catalogFramework.delete(new DeleteRequestImpl(id));
+
           subscriptions.removeSubscriptions(id);
           return ImmutableMap.of("message", "Successfully deleted.");
         },
         util::getJson);
 
+    get(
+        "/workspace/:id/queries",
+        (req, res) -> {
+          String workspaceId = req.params(":id");
+          WorkspaceMetacardImpl workspace = workspaceService.getWorkspaceMetacard(workspaceId);
+
+          List<String> queryIds = workspace.getQueries();
+
+          return util.getMetacardsWithTagById(queryIds, QUERY_TAG)
+              .values()
+              .stream()
+              .map(Result::getMetacard)
+              .map(queryTransformer::transformMetacardIntoMap)
+              .collect(Collectors.toList());
+        },
+        util::getJson);
+
+    post(
+        "/query",
+        APPLICATION_JSON,
+        (req, res) -> {
+          Map<String, Object> body = mapper.parser().parseMap(util.safeGetBody(req));
+
+          Metacard metacard = queryTransformer.transformJsonIntoMetacard(body);
+          Metacard saved = saveMetacard(metacard);
+
+          res.status(201);
+          return queryTransformer.transformMetacardIntoMap(saved);
+        },
+        util::getJson);
+
+    put(
+        "/query/:id",
+        APPLICATION_JSON,
+        (req, res) -> {
+          String queryId = req.params("id");
+          Map<String, Object> body = mapper.parser().parseMap(util.safeGetBody(req));
+
+          Metacard metacard = queryTransformer.transformJsonIntoMetacard(body);
+          Metacard updated = updateMetacard(queryId, metacard);
+
+          return queryTransformer.transformMetacardIntoMap(updated);
+        },
+        util::getJson);
+
+    get(
+        "/query/:id",
+        (req, res) -> {
+          String id = req.params("id");
+          Metacard metacard = util.getMetacardById(id);
+
+          Set<String> metacardTags = metacard.getTags();
+
+          if (metacardTags == null || !metacardTags.contains(QUERY_TAG)) {
```

Comment:
Check for apache collections method that performs null check on set

Reply:
I found a couple of different apache commons libraries that perform operations like this. However, they all require the set(s) to be non-null.